subsys/modbus/mb_rtu_client.c

```diff
@@ -0,0 +1,615 @@
+		if (IS_ENABLED(CONFIG_MODBUS_RTU_FP_EXTENSIONS) &&
+		    (req_addr >= MODBUS_RTU_FP_ADDR)) {
+			err = mbm_validate_fc03fp_response(pch, (float *)data);
+			break;
+		}
+		/* fallthrough */
+	case MODBUS_FC04_IN_REG_RD:
+		req_byte_cnt = req_qty * sizeof(uint16_t);
+		if (req_byte_cnt != resp_byte_cnt) {
+			LOG_ERR("Mismatch in the number of registers");
+			err = -EINVAL;
+		} else {
+			for (uint8_t i = 0; i < req_qty; i++) {
+				data_p16[i] = sys_get_be16(resp_data);
+				resp_data += sizeof(uint16_t);
+			}
+
+			err = 0;
+		}
+		break;
+
+	default:
+		LOG_ERR("Validation not implemented for FC 0x%02x", fc);
+		err = -ENOTSUP;
+	}
+
+	return err;
+}
+
+static int mbm_validate_fc08_response(struct mb_rtu_context *pch,
+				      const uint8_t node_addr,
+				      uint16_t *data)
+{
+	int err;
+	uint16_t resp_sfunc;
+	uint16_t resp_data;
+	uint16_t req_sfunc;
+	uint16_t req_data;
+
+	if (data == NULL) {
+		return -EINVAL;
+	}
+
+	req_sfunc = sys_get_be16(&pch->tx_frame.data[0]);
+	req_data = sys_get_be16(&pch->tx_frame.data[2]);
+	resp_sfunc = sys_get_be16(&pch->rx_frame.data[0]);
+	resp_data = sys_get_be16(&pch->rx_frame.data[2]);
+
+	if (req_sfunc != resp_sfunc) {
+		LOG_ERR("Mismatch in the sub-function code");
+		return -EINVAL;
+	}
+
+	switch (resp_sfunc) {
+	case MODBUS_FC08_SUBF_QUERY:
+	case MODBUS_FC08_SUBF_CLR_CTR:
+		if (req_data != resp_data) {
+			LOG_ERR("Request and response data are different");
+			err = -EINVAL;
+		} else {
+			*data = resp_data;
+			err = 0;
+		}
+		break;
+
+	case MODBUS_FC08_SUBF_BUS_MSG_CTR:
+	case MODBUS_FC08_SUBF_BUS_CRC_CTR:
+	case MODBUS_FC08_SUBF_BUS_EXCEPT_CTR:
+	case MODBUS_FC08_SUBF_SERVER_MSG_CTR:
+	case MODBUS_FC08_SUBF_SERVER_NO_RESP_CTR:
+		*data = resp_data;
+		err = 0;
+		break;
+
+	default:
+		err = -EINVAL;
+	}
+
+	return err;
+}
+
+static int mbm_validate_wr_response(struct mb_rtu_context *pch,
+				    const uint8_t node_addr,
+				    uint8_t fc)
+{
+	int err;
+	uint16_t req_addr;
+	uint16_t req_value;
+	uint16_t resp_addr;
+	uint16_t resp_value;
+
+	req_addr = sys_get_be16(&pch->tx_frame.data[0]);
+	req_value = sys_get_be16(&pch->tx_frame.data[2]);
+	resp_addr = sys_get_be16(&pch->rx_frame.data[0]);
+	resp_value = sys_get_be16(&pch->rx_frame.data[2]);
+
+	switch (fc) {
+	case MODBUS_FC05_COIL_WR:
+	case MODBUS_FC06_HOLDING_REG_WR:
+	case MODBUS_FC15_COILS_WR:
+	case MODBUS_FC16_HOLDING_REGS_WR:
+		if (req_addr != resp_addr || req_value != resp_value) {
+			err = ENXIO;
+		} else {
+			err = 0;
+		}
+		break;
+
+	default:
+		LOG_ERR("Validation not implemented for FC 0x%02x", fc);
+		err = -ENOTSUP;
+	}
+
+	return err;
+}
+
+static int mbm_send_cmd(struct mb_rtu_context *ctx, const uint8_t node_addr,
+			uint8_t fc, void *data)
+{
+	int err;
+
+	ctx->tx_frame.addr = node_addr;
+	ctx->tx_frame.fc = fc;
+
+	mb_tx_frame(ctx);
+
+	if (k_sem_take(&ctx->client_wait_sem, K_USEC(ctx->rxwait_to)) != 0) {
+		LOG_WRN("Client wait-for-RX timeout");
+		err = -EIO;
+		goto exit_error;
```

Comment:
if there is no need for cleanup on error, direct return may be used instead of `goto`

Reply:
fixed